GAE/src/org/waterforpeople/mapping/dao/SurveyInstanceDAO.java

```diff
@@ -99,21 +99,17 @@ public SurveyInstance save(SurveyInstance si, DeviceFiles deviceFile) {
         final Set<QuestionAnswerStore> locations = new HashSet<>();
         final List<QuestionAnswerStore> responses = new ArrayList<>();
         for (QuestionAnswerStore qas : si.getQuestionAnswersStore()) {
-            qas.setSurveyInstanceId(surveyInstanceId);
-            if (qasDao.isCached(Long.valueOf(qas.getQuestionID()), surveyInstanceId)) {
-                log.log(Level.INFO,
-                        "Skipping QAS already present in datasore [SurveyInstance, Survey, Question]: "
-                                + surveyInstanceId + ", " + si.getSurveyId() + ", " + qas.getQuestionID());
-                continue; // Already stored
-            }
-            
-            if (Question.Type.GEO.toString().equals(qas.getType()) && isNew) {
-                locations.add(qas);
-            } else if ("IMAGE".equals(qas.getType())) {
-                // The device send values as IMAGE and not PHOTO.
-                images.add(qas);
+            if (isValid(qas, si)) {
```

Comment:
I'd argue `saveResponse(Response response)` kind of suggests side effects (saving the response), rather than merely checking whether the response is valid or not.